libs/ngx-sherfire/src/lib/firestore/derivable-firestore.service.ts

```diff
@@ -0,0 +1,136 @@
+import { Derivable, DerivableAtom, SettableDerivable, error, lens } from '@skunkteam/sherlock';
+import { DerivableCacheOptions, derivableCache, fromEventPattern } from '@skunkteam/sherlock-utils';
+import {
+    CollectionReference,
+    DocumentData,
+    DocumentReference,
+    DocumentSnapshot,
+    FirestoreError,
+    Query,
+    QueryDocumentSnapshot,
+    QuerySnapshot,
+    onSnapshot,
+    setDoc,
+} from 'firebase/firestore';
+import { NgxSherfireModule } from '../ngx-sherfire.module';
+
+@Injectable({ providedIn: NgxSherfireModule })
+export class DerivableFirestore {
+    constructor(private readonly zone: NgZone) {}
+
+    /**
+     * Build a derivable based on a Query or CollectionReference that returns a stream of updated QuerySnapshots that contain all
+     * QueryDocumentSnapshots matching the query.
+     */
+    snapshot$<App, Db extends DocumentData>(query: Query<App, Db>): Derivable<QuerySnapshot<App, Db>>;
+    /**
+     * Build a derivable based on a DocumentReference that returns a stream of updated DocumentSnapshots.
+     */
+    snapshot$<App, Db extends DocumentData>(ref: DocumentReference<App, Db>): Derivable<DocumentSnapshot<App, Db>>;
+    snapshot$<App, Db extends DocumentData>(
+        v: DocumentReference<App, Db> | Query<App, Db>,
+    ): Derivable<DocumentSnapshot<App, Db>> | Derivable<QuerySnapshot<App, Db>> {
+        return v instanceof DocumentReference
+            ? this.CACHES.doc.snapshot(v)
+            : v instanceof CollectionReference
+            ? this.CACHES.collection.snapshot(v)
+            : this.querySnapshot$(v);
+    }
+
+    /**
+     * Build a derivable based on a Query or CollectionReference that returns a stream of updates with an array of QueryDocumentSnapshot.
+     */
+    docs$<App, Db extends DocumentData>(query: Query<App, Db>): Derivable<Array<QueryDocumentSnapshot<App, Db>>> {
+        return query instanceof CollectionReference ? this.CACHES.collection.docs(query) : this.queryDocs$(query);
+    }
+
+    /**
+     * Build a derivable based on a Query or CollectionReference that returns a stream of arrays of document contents.
+     */
+    data$<App, Db extends DocumentData>(query: Query<App, Db>): Derivable<App[]>;
+    /**
+     * Build a derivable based on a DocumentReference that returns a stream of document contents. The returned derivable is settable and
+     * setting a value to the derivable will update the firestore document.
+     */
+    data$<App, Db extends DocumentData>(ref: DocumentReference<App, Db>): SettableDerivable<App>;
+    data$<App, Db extends DocumentData>(
+        v: DocumentReference<App, Db> | Query<App, Db>,
+    ): SettableDerivable<App> | Derivable<App[]> {
+        return v instanceof DocumentReference
+            ? this.CACHES.doc.data(v)
+            : v instanceof CollectionReference
+            ? this.CACHES.collection.data(v)
+            : this.queryData$(v);
+    }
+
+    private readonly CACHES = this.buildCaches();
```

Comment:
I think we should consider removing the cache. It is currently not sound. It is possible to use different converters on your query or doc-reference, which is not reflected in the `path` property. So, either we should take that into account, or we should simply remove the cache and leave it to the consumer to cache this if needed. I believe the last option to be best.

Reply:
Good point. We also discussed that a user might previously use `{{service.docRef.data$}}` in an angular template and this would need the derivableCache to prevent creating new derivables all the time. But you now have to invoke `firestoreService.data$(docRef)` and people should know not to do this from templates but from their service and easily cache ther derivable there.